components/core/src/clp/ffi/ir_stream/Serializer.cpp

```diff
@@ -59,6 +258,207 @@ auto Serializer<encoded_variable_t>::change_utc_offset(UtcOffset utc_offset) ->
+            auto const& inner_map{val.via.map};
+            auto const inner_map_size(static_cast<size_t>(inner_map.size));
+            if (0 == inner_map_size) {
+                if (false == serialize_key(schema_tree_node_id)) {
+                    failure = true;
+                    break;
+                }
+                serialize_empty_object(m_value_group_buf);
+            } else {
+                working_stack.emplace_back(schema_tree_node_id, inner_map.ptr, inner_map_size);
+            }
+        } else {
+            // Serialize primitive values
+            if (false
+                == (serialize_key(schema_tree_node_id) && serialize_val(val, schema_tree_node_type)
+                ))
+            {
+                failure = true;
+                break;
+            }
+        }
+    }
+
+    if (failure) {
+        m_schema_tree.revert();
+        return false;
+    }
+
+    m_ir_buf.insert(
+            m_ir_buf.cend(),
+            m_schema_tree_node_buf.cbegin(),
+            m_schema_tree_node_buf.cend()
+    );
+    m_ir_buf.insert(m_ir_buf.cend(), m_key_group_buf.cbegin(), m_key_group_buf.cend());
+    m_ir_buf.insert(m_ir_buf.cend(), m_value_group_buf.cbegin(), m_value_group_buf.cend());
+    return true;
+}
+
+template <typename encoded_variable_t>
+auto Serializer<encoded_variable_t>::serialize_schema_tree_node(
+        SchemaTree::NodeLocator const& locator
+) -> bool {
+    switch (locator.get_type()) {
+        case SchemaTreeNode::Type::Int:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeInt);
+            break;
+        case SchemaTreeNode::Type::Float:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeFloat);
+            break;
+        case SchemaTreeNode::Type::Bool:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeBool);
+            break;
+        case SchemaTreeNode::Type::Str:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeStr);
+            break;
+        case SchemaTreeNode::Type::UnstructuredArray:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeUnstructuredArray);
+            break;
+        case SchemaTreeNode::Type::Obj:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeObj);
+            break;
+        default:
+            // Unknown type
+            return false;
+    }
+    auto const parent_id{locator.get_parent_id()};
+    if (parent_id <= UINT8_MAX) {
+        m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeParentIdUByte);
+        m_schema_tree_node_buf.push_back(static_cast<int8_t>(static_cast<uint8_t>(parent_id)));
+    } else if (parent_id <= UINT16_MAX) {
+        m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeParentIdUShort);
+        serialize_int(static_cast<uint16_t>(parent_id), m_schema_tree_node_buf);
+    } else {
```

Comment:
I guess it's fine for now, but we'll probably want to support 32 bit node ids to feel comfortable. I guess the current strategy is to split the ir stream if we pass this threshold? Should be fine, but there are definitely some datasets that can slip past 64K within several GB of data.

Reply:
If we want to add support for larger schema tree, we just need to add a header byte to indicate 32bit encoded node ids. The header byte is reserved